keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -526,302 +517,382 @@ dualstack cluster can have this flag configured as
+  preferDualStack: true # set by user
+```
 
-# dual stack service (optional)
+...would produce the following result on a single stack cluster:
+
+```yaml
 apiVersion: v1
 kind: Service
 metadata:
   name: my-service
 spec:
   type: ClusterIP
-  preferDualStack: true # set by user
-  ipFamilies:
-    - IPv6 # defaulted, cluster does not support dual stack services, only one IPv6 CIDR is provided for services
-  clusterIP: 2001::1  # allocated
-  clusterIPs: # allocated
-    - 2001::1 # note single entry
   selector:
     app: MyApp
   ports:
     - protocol: TCP
       port: 80
       targetPort: 9376
+  preferDualStack: true # set by user
+  ipFamilies:           # defaulted
+    - IPv6              # note single entry
+  clusterIP: 2001::1    # allocated
+  clusterIPs:           # allocated
+    - 2001::1           # note single entry
+```
+
+On a dual stack cluster it would have produced:
 
-# dual stack service (optional)
+```yaml
 apiVersion: v1
 kind: Service
 metadata:
   name: my-service
 spec:
   type: ClusterIP
-  preferDualStack: true # set by user
-  ipFamilies:
-    - IPv6 # set by user
-    - IPv4 # defaulted (cluster supports dual stack)
-  clusterIP: 2001::1  # allocated
-  clusterIPs: # allocated
-    - 2001::1
-    - 1.2.3.4
   selector:
     app: MyApp
   ports:
     - protocol: TCP
       port: 80
       targetPort: 9376
-
+  preferDualStack: true # set by user
+  ipFamilies:           # defaulted
+    - IPv6              # note two entries
+    - IPv4 
+  clusterIP: 2001::1    # allocated
+  clusterIPs:           # allocated
+    - 2001::1           # note two entries
+    - 1.2.3.4
 ```
 
 ##### Notes on NodePort Allocation
 
-NodePort reservation will be constant across both families. That mean a
-reservation of `NodePort: 12345` will happen across both families irrespective
-of the IP family of the service. `kube-proxy` will ensure that traffic is
+NodePort reservations will apply across both families. That means a
+reservation of `NodePort: 12345` will happen on both families, irrespective
+of the IP family of the service. A single service which is dual stack may use
+the same NodePort on both families.  `kube-proxy` will ensure that traffic is
 routed according to the families assigned for services.
 
 #### Type Headless services
 
-1. For Headless with selector, the value of `spec.preferDualStack` and
-   `spec.ipFamilies` will drive the endpoint selection. Endpoint slice
-   controller will select endpoints with IP families that match the value of
-   these two flags.
+1. For Headless with selector, the value of `spec.ipFamilies` will drive the
+   endpoint selection. Endpoints will use the value of `ipFamiles[0]`.
+   EndpointSlices will use endpoints with all IP families of the Service.
 2. For Headless without selector the values of `spec.preferDualStack` and
-   `spec.ipFamilies` are treated as saved as they were provided by the user.
+   `spec.ipFamilies` are saved as they were provided by the user.
```

Comment:
That implies they are not defaulted, which I don't think was meant. "For Headless without selector, the defaulted or user-provided values of ... are saved but have no effect" ?

(Is that correct that they have no effect? A selectorless headless Service with single-stack `ipFamilies` but dual-stack `Endpoints` is expected to get dual-stack DNS, right?)

Reply:
The latter is a great question.

@johnbelamaric @chrisohaver - if I have a Service that has no selector (manually configured) and has no VIP (headless) AND has Endpoints with both IPv4 and IPv6, what do we expect CoreDNS to do?  

I would expect it to follow the new `ipFamiles` field, but that could represent a regression vs today.  What happens today?

Reply:
I see empirically that CoreDNS programs both A and AAAA in this case today, which means that if we demand that it follows `ipFamilies` it is a functional break.  If we don't demand that it follows `ipFamilies` it is inconsistent and confusing.

Thoughts?

Reply:
If we want to keep the current A+AAAA behavior without having inconsistency, we could say that headless selectorless Services, like ExternalName Services, are "expected" to have `ipFamilies: nil` (for whatever definition of "expected" we decide on below).

Reply:
Do you think we should add a dependency on ipFamily to the DNS?
I think that is fine right now, DNS create records for the endpoints it gets, we can control the DNS records from the endpoints objects.

Reply:
Dan and Kal and I talked, and we have a proposal for these corner cases, which will come in a followup PR